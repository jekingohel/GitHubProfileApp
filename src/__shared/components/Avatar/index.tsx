import React from 'react';
import {Image, Pressable} from 'react-native';
import Text from '../Text';

interface AvatarProps {
  source: string | null;
  title: string;
  rounded?: boolean;
  size: 'large' | 'medium' | 'small';
  onPress?: () => void;
  backgroundColor?: string;
  textColor?: string;
  customColor?: string | null;
  customSize?: number;
  loading?: boolean;
  containerStyle?: object;
  singleLetter?: boolean;
  fullView?: boolean;
  imageProps?: object;
}

const sizeArray = {
  small: 28,
  medium: 38,
  large: 70,
};

const Avatar: React.FC<AvatarProps> = props => {
  const {
    containerStyle,
    singleLetter,
    fullView,
    imageProps = {},
    source = null,
    title = '',
    rounded = false,
    size = 'medium',
    onPress = () => {},
    backgroundColor = 'rgba(142,142,147,.8)',
    textColor = '#FFFFFF',
  } = props;
  const initialis = title
    ?.match(/\b(\w)/g)
    ?.slice(0, 2)
    ?.join('')
    ?.toUpperCase();

  return source === null ? (
    <Pressable
      onPress={onPress}
      disabled={!onPress}
      style={[
        {
          width: sizeArray[size],
          height: sizeArray[size],
          backgroundColor,
          alignItems: 'center',
          justifyContent: 'center',
          borderRadius: rounded ? sizeArray[size] : 0,
        },
        containerStyle && containerStyle,
      ]}
      {...props}>
      <Text
        color={textColor}
        fontWeight="400"
        style={{alignSelf: 'center'}}
        fontSize={sizeArray[size] / 2.2}
        lineHeight={sizeArray[size] + sizeArray[size] / 10}>
        {singleLetter ? title?.charAt(0)?.toUpperCase() : initialis}
      </Text>
    </Pressable>
  ) : (
    <Pressable
      {...props}
      disabled={!onPress && !fullView}
      style={[{position: 'relative'}]}>
      <Image
        {...imageProps}
        source={{uri: source}}
        style={[
          {
            borderRadius: rounded ? sizeArray[size] : 0,
            width: sizeArray[size],
            height: sizeArray[size],
          },
          containerStyle && containerStyle,
        ]}
      />
    </Pressable>
  );
};

export default Avatar;
