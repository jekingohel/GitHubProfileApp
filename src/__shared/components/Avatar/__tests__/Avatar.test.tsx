import React from 'react';
import {render, fireEvent} from '@testing-library/react-native';
import Avatar from '..';

test('renders Avatar with a single letter correctly', () => {
  const {getByText, toJSON} = render(
    <Avatar source={null} title="John Doe" size="medium" />,
  );
  const avatarText = getByText('JD'); // Assumes "John Doe" renders as "JD"
  expect(avatarText).toBeTruthy();
  expect(toJSON()).toMatchSnapshot();
});

test('renders Avatar with an image correctly', () => {
  const {getByTestId, toJSON} = render(
    <Avatar
      title="Test Case"
      source="https://dummyimage.com/100"
      size="large"
    />,
  );
  const avatarImage = getByTestId('avatar-image');
  expect(avatarImage).toBeTruthy();
  expect(toJSON()).toMatchSnapshot();
});

test('handles onPress event correctly', () => {
  const onPressMock = jest.fn();
  const {getByTestId} = render(
    <Avatar
      source="https://dummyimage.com/100"
      title="Jane Doe"
      size="small"
      onPress={onPressMock}
    />,
  );
  const avatarPressable = getByTestId('avatar-pressable');
  fireEvent.press(avatarPressable);
  expect(onPressMock).toHaveBeenCalled();
});

test('renders Avatar with custom size and color correctly', () => {
  const {toJSON} = render(
    <Avatar
      source="https://dummyimage.com/100"
      title="Alice"
      size="medium"
      customSize={10}
      backgroundColor="#ff0000"
      textColor="#ffffff"
    />,
  );
  expect(toJSON()).toMatchSnapshot();
});
