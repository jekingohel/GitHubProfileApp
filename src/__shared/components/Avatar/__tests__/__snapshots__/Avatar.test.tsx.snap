// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders Avatar with a single letter correctly 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": false,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  size="medium"
  source={null}
  style={
    [
      {
        "alignItems": "center",
        "backgroundColor": "rgba(142,142,147,.8)",
        "borderRadius": 0,
        "height": 38,
        "justifyContent": "center",
        "width": 38,
      },
      undefined,
    ]
  }
  title="John Doe"
>
  <Text
    style={
      {
        "alignSelf": "center",
        "color": "#FFFFFF",
        "fontSize": 17.27272727272727,
        "fontStyle": "normal",
        "fontWeight": "400",
        "lineHeight": 41.8,
        "textAlign": "left",
      }
    }
  >
    JD
  </Text>
</View>
`;

exports[`renders Avatar with an image correctly 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": false,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  size="large"
  source="https://dummyimage.com/100"
  style={
    [
      {
        "position": "relative",
      },
    ]
  }
  testID="avatar-pressable"
  title="Test Case"
>
  <Image
    source={
      {
        "uri": "https://dummyimage.com/100",
      }
    }
    style={
      [
        {
          "borderRadius": 0,
          "height": 65,
          "width": 65,
        },
        undefined,
      ]
    }
    testID="avatar-image"
  />
</View>
`;

exports[`renders Avatar with custom size and color correctly 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": false,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  backgroundColor="#ff0000"
  collapsable={false}
  customSize={10}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  size="medium"
  source="https://dummyimage.com/100"
  style={
    [
      {
        "position": "relative",
      },
    ]
  }
  testID="avatar-pressable"
  textColor="#ffffff"
  title="Alice"
>
  <Image
    source={
      {
        "uri": "https://dummyimage.com/100",
      }
    }
    style={
      [
        {
          "borderRadius": 0,
          "height": 48,
          "width": 48,
        },
        undefined,
      ]
    }
    testID="avatar-image"
  />
</View>
`;
